// @decorators/express/lib/src/decorators/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const meta_1 = require("../meta");
/**
 * Parameter decorator factory, creates parameter decorator
 *
 * @param {ParameterType} parameterType Parameter Type
 */
function decoratorFactory(type) {
    return function (name) {
        return function (target, methodName, index) {
            const meta = meta_1.getMeta(target);
            if (meta.params[methodName] === undefined) {
                meta.params[methodName] = [];
            }
            meta.params[methodName].push({ index, type, name });
        };
    };
}
/**
 * Express req object
 */
exports.Request = decoratorFactory(meta_1.ParameterType.REQUEST);
/**
 * Express res object
 */
exports.Response = decoratorFactory(meta_1.ParameterType.RESPONSE);
/**
 * Express next function
 */
exports.Next = decoratorFactory(meta_1.ParameterType.NEXT);
/**
 * Express req.params object or single param, if param name was specified
 */
exports.Params = decoratorFactory(meta_1.ParameterType.PARAMS);
/**
 * Express req.query object or single query param, if query param name was specified
 */
exports.Query = decoratorFactory(meta_1.ParameterType.QUERY);
/**
 * Express req.body object or single body param, if body param name was specified
 */
exports.Body = decoratorFactory(meta_1.ParameterType.BODY);
/**
 * Express req.headers object or single headers param, if headers param name was specified
 */
exports.Headers = decoratorFactory(meta_1.ParameterType.HEADERS);
/**
 * Express req.body object or single cookies param, if cookies param name was specified
 */
exports.Cookies = decoratorFactory(meta_1.ParameterType.COOKIES);
