// @decorators/express/lib/src/decorators/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const meta_1 = require("../meta");
/**
 * Route decorator factory, creates decorator
 *
 * @param {string} method
 * @param {string} url
 * @param {Type[]} middleware
 */
function decoratorFactory(method, url, middleware) {
    return (target, key, descriptor) => {
        const meta = meta_1.getMeta(target);
        meta.routes[key] = { method, url, middleware };
        return descriptor;
    };
}
/**
 * All routes
 *
 * Special-cased "all" method, applying the given route `path`,
 * middleware, and callback to _every_ HTTP method.
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function All(url, middleware) {
    return decoratorFactory('all', url, middleware);
}
exports.All = All;
;
/**
 * Get route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Get(url, middleware) {
    return decoratorFactory('get', url, middleware);
}
exports.Get = Get;
;
/**
 * Post route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Post(url, middleware) {
    return decoratorFactory('post', url, middleware);
}
exports.Post = Post;
/**
 * Put route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Put(url, middleware) {
    return decoratorFactory('put', url, middleware);
}
exports.Put = Put;
/**
 * Delete route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Delete(url, middleware) {
    return decoratorFactory('delete', url, middleware);
}
exports.Delete = Delete;
/**
 * Patch route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Patch(url, middleware) {
    return decoratorFactory('patch', url, middleware);
}
exports.Patch = Patch;
/**
 * Options route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Options(url, middleware) {
    return decoratorFactory('options', url, middleware);
}
exports.Options = Options;
/**
 * Head route
 *
 * @param {string} url
 * @param {Type[]} [middleware]
 */
function Head(url, middleware) {
    return decoratorFactory('head', url, middleware);
}
exports.Head = Head;
